crates/rpc/rpc/src/eth/api/server.rs

```diff
@@ -180,11 +181,84 @@ where
 
     async fn fee_history(
         &self,
-        _block_count: U256,
-        _newest_block: BlockNumber,
+        block_count: u64,
+        newest_block: BlockNumber,
         _reward_percentiles: Option<Vec<f64>>,
     ) -> Result<FeeHistory> {
-        Err(internal_rpc_err("unimplemented"))
+        if block_count == 0 || newest_block < block_count {
+            return Ok(FeeHistory::default())
+        }
+
+        let start_block = newest_block - block_count;
+        let end_block = newest_block;
+
+        let mut fee_history_cache = self.fee_history_cache.lock();
+
+        let mut fee_history_cache_items = BTreeMap::new();
+        let mut first_non_cached_block = None;
+        for block in start_block..=end_block {
+            // Check if block exists in cache, and move it to the head of the list if so
+            if let Some(fee_history_cache_item) = fee_history_cache.get(&block) {
+                fee_history_cache_items.insert(block, fee_history_cache_item.clone());
+            } else {
+                // If block doesn't exist in cache, set it as a first non-cached block to query it
+                // from the database and break the loop since we'll overwrite the remaining blocks
+                // in cache anyway.
+                first_non_cached_block.get_or_insert(block);
+                break
+            }
+        }
+
+        // If we had any cache misses, query the database starting with the first non-cached block
+        if let Some(start_block) = first_non_cached_block {
+            let headers: Vec<Header> =
+                self.inner.client.headers_range(start_block..=end_block).to_rpc_result()?;
+
+            // We should receive exactly the amount of blocks missing from the cache
+            if headers.len() != (end_block - start_block + 1) as usize {
+                return Ok(FeeHistory::default())
+            }
+
+            for header in headers {
+                let base_fee_per_gas = header.base_fee_per_gas.
+                        unwrap_or_default(). // Zero for pre-EIP-1559 blocks
+                        try_into().unwrap(); // u64 -> U256 won't fail
+                let gas_used_ratio = header.gas_used as f64 / header.gas_limit as f64;
+
+                let fee_history_cache_item = FeeHistoryCacheItem {
+                    hash: None,
+                    base_fee_per_gas,
+                    gas_used_ratio,
+                    reward: None, // TODO: calculate rewards per transaction
+                };
+
+                fee_history_cache_items.insert(header.number, fee_history_cache_item.clone());
+                fee_history_cache.push(header.number, fee_history_cache_item);
```

Comment:
is this right?
I don't thin the cache is supposed to be populated on an eth_feeHistory call?

Reply:
this is exactly what happens in geth, if I understood it right:
https://github.com/ethereum/go-ethereum/blob/77380b9629e4f679778ce9de037ffdaf9c925238/eth/api_backend.go#L305
https://github.com/ethereum/go-ethereum/blob/77380b9629e4f679778ce9de037ffdaf9c925238/eth/gasprice/feehistory.go#L296

we can re-evaluate whether it's better to update the cache on every block execution as in Anvil, or to do it lazily on `eth_feeHistory` RPC calls. Performance-wise, I think it's better to have cache warmed on block execution, but some users may not need `eth_feeHistory` RPC, so for them it'll be a tiny bit increased block execution + memory footprint for nothing.

Reply:
oh so this is lazy eval, got it. this is fine for now imo.

please document this.

Reply:
documented